bpo-26253: Add compressionlevel to tarfile stream

jarondl opened the pull request:
`tarfile` already accepts a compressionlevel argument for creating
files. This patch adds the same for stream-based tarfile usage.
The default is 9, the value that was previously hard-coded.

<!-- issue-number: bpo-26253 -->
https://bugs.python.org/issue26253
<!-- /issue-number -->

#70441

jarondl committed:
bpo-26253: Add compressionlevel to tarfile stream

`tarfile` already accepts a compressionlevel argument for creating
files. This patch adds the same for stream-based tarfile usage.
The default is 9, the value that was previously hard-coded.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

serhiy-storchaka commented on Lib/tarfile.py:
```diff
@@ -357,14 +358,15 @@ def __init__(self, name, mode, comptype, fileobj, bufsize):
-        self.closed   = False
+        self.fileobj = fileobj
+        self.bufsize = bufsize
+        self.compresslevel = compresslevel
```
Is it needed to save `compresslevel` as an attribute? Wouldn't be enough to pass `compresslevel` as an argument to `_init_write_gz()`?

jarondl replied:
No it is not needed. My first approach was to pass it as an argument to `_init_write_gz()`, with a default value of 9 (not to change the signature).  But then I got the sense that an attribute goes more naturally with the rest of the class. I can revert to the first approach if you prefer.

serhiy-storchaka replied:
Other attributes are used in multiple `_Stream` methods or are needed for simulating a file object. But `compresslevel` is used only in the constructor (`_init_write_gz()` is a factored out part of the constructor). It is not needed after creating a `_Stream` object.

serhiy-storchaka commented on Doc/library/tarfile.rst:
```diff
@@ -98,8 +98,8 @@ Some facts and figures:
 
-   For modes ``'w:gz'``, ``'r:gz'``, ``'w:bz2'``, ``'r:bz2'``, ``'x:gz'``,
-   ``'x:bz2'``, :func:`tarfile.open` accepts the keyword argument
+   For modes ``'w:gz'``, ``'x:gz'``, ``'w|gz'``, ``'w:bz2'``, ``'x:bz2'``,
```
Add the `versionchanged` directive.

Document the change in the What's New document.

jarondl replied:
`versionchanged` - Done

serhiy-storchaka reviewed (commented):
New feature needs new tests.

serhiy-storchaka commented on Lib/tarfile.py:
```diff
@@ -343,7 +343,8 @@ class _Stream:
 
-    def __init__(self, name, mode, comptype, fileobj, bufsize):
+    def __init__(self, name, mode, comptype, fileobj, bufsize,
+                 compresslevel=9):
```
Since `_Stream` is intended to be used only internally, we are free to change the signature of its constructor. The default argument for `compresslevel` is not needed. This just duplicates its value.

jarondl committed:
bpo-26253: don't add compresslevel as an attribute

In tarfile, compresslevel will be simply passed around.

Also added 'versionadded' to stream compression level setting

jarondl commented:
Thank you @serhiy-storchaka for the review. I have changed the attribute, and added a `versionschanged`. However, I don't think this is a whatsnew material - it just makes the current `compresslevel` feature more consistent. The bug report is about unexpected results, not missing features. But perhaps I am wrong about that.

Regarding testing, there are currently no tests for compression at all. I am thinking of adding tests that will check argument acceptance / rejection (i.e. raise an error for `compresslevel` with `w:` etc)
Maybe also write to a tarfile with compression and check the first few bytes to see the compression signature. Checking actual values or sizes is problematic because of build constants

jarondl committed:
bpo-26253: tarfile: add compresslevel unit tests

Check outcoming file headers when writing a compressed file.

jarondl commented:
tests added

MaxwellDupre reviewed (approved):
Ran test_tarfile with:
Ran 454 tests in 28.170s
OK (skipped=8)
Looks ok, but noted this was against 3.7 build.

AA-Turner commented on Doc/library/tarfile.rst:
```diff
@@ -152,6 +152,9 @@ Some facts and figures:
    .. versionchanged:: 3.6
       The *name* parameter accepts a :term:`path-like object`.
 
+   .. versionchanged:: 3.7
```
```suggestion
   .. versionchanged:: 3.12
```

AA-Turner commented on Misc/NEWS.d/next/Library/2017-07-31-13-35-28.bpo-26253.8v_sCs.rst:
```diff
@@ -0,0 +1 @@
+Adjustable compression level for tarfile streams.
```
```suggestion
Allow adjustable compression level for tarfile streams in 
:function:`tarfile.open`.
```

AA-Turner replied:
Could you also update the NEWS file to point to #70441 rather than b.p.o -- you may need to regenerate the file, or rename it to the new pattern.

A

AA-Turner reviewed (approved):
Looks OK barring two documentation items:

serhiy-storchaka committed:
Apply suggestions from code review

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

AA-Turner commented on Misc/NEWS.d/next/Library/2017-07-31-13-35-28.bpo-26253.8v_sCs.rst:
```diff
@@ -0,0 +1,2 @@
+Allow adjustable compression level for tarfile streams in 
+:function:`tarfile.open`.
```
CI failing due to my suggestion; @serhiy-storchaka this should fix it:
```suggestion
Allow adjustable compression level for tarfile streams in
:func:`tarfile.open`.
```

A

serhiy-storchaka committed:
Update Misc/NEWS.d/next/Library/2017-07-31-13-35-28.bpo-26253.8v_sCs.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>
